File: examples/nodejs/level-1-streaming.js
const ws = require('ws');

function parseLevel1(message) {
    let dataArray = message.split('~');
    let dataObj = {
        type: dataArray[0],
        market: dataArray[1],
        fromsymbol: dataArray[2],
        tosymbol: dataArray[3],
        side: dataArray[4] === '1' ? 'bid' : 'ask',
        sequence: parseInt(dataArray[5], 10),
        price: parseFloat(dataArray[6]),
        quantity: parseFloat(dataArray[7]),
        timestamp: new Date(parseInt(dataArray[8],10)/1000000)
    }
    return dataObj;
}


let cryptocompareWS = new ws('wss://streaming.cryptocompare.com');

let subs = {
    "action": "SubAdd",
    "subs": ["30~bitfinex~BTC~USD"],
    "api_key": "YOUR-API-KEY",
    "format": "streamer"
}

cryptocompareWS.on('open', function(){
    console.log('connection established')
    cryptocompareWS.send(JSON.stringify(subs));
});

cryptocompareWS.on('error', function(error){
    console.log(error)
});

cryptocompareWS.on('close', function(){
    console.log('disconnected');
});

cryptocompareWS.on('message', function(data){
    console.log(data)
    if (data === '999~HEARTBEAT|') {
        console.log(data);
        return;
    }
    let updates = data.split('|');
    updates.pop(); //Remove last element as it is always an empty string
    for (let update of updates) {
        let parsedUpdate = parseLevel1(update);
        console.log(JSON.stringify(parsedUpdate))
    }
});

File: examples/python/level-1-streaming.py
from lomond import WebSocket
from lomond.persist import persist

def parse_level_1(message):
    data_array = message.split('~');
    data_obj = {
        "message_type": data_array[0],
        "exchange": data_array[1],
        "fromsymbol": data_array[2],
        "tosymbol": data_array[3],
        "side": 'bid' if data_array[4] == '1' else 'ask',
        "sequence": int(data_array[5]),
        "price": float(data_array[6]),
        "quantity": float(data_array[7]),
        "timestamp": int(data_array[8])/1000000
    }
    return data_obj

cryptocompare_ws = WebSocket('wss://streaming.cryptocompare.com')
subs = {
    "action": "SubAdd",
    "subs": ["30~bitfinex~BTC~USD"],
    "api_key": "YOUR-API-KEY",
    "format": "streamer"
}

for event in persist(cryptocompare_ws):
    if event.name == "ready":
        print('ready')
        cryptocompare_ws.send_json(subs)
    if event.name == "text":
        messages = event.text.split("|")
        messages = messages[:-1] #Last element is empty string
        for message in messages:
            print(parse_level_1(message))